Initial implementation of the Snapshot API

wecc opened the pull request:
This is a first stab at the new Snapshot API, `serializer.serialize()` now receives a Snapshot instead of a record instance.

- **Caching** - Right now, attributes are cached eagerly (on instantiation) and relationships are cached on first access. Is this the design we want?
- **snapshot.belongsTo()** - Returns a new snapshot (with at least an ID) or undefined. Is this what we want or should we always return a new snapshot?
- **snapshot.hasMany()** - Always returns an array.
- **snapshot.get() deprecated** - ~~To only use `attr()`, `belongsTo()` and `hasMany()` would require some work in serializers. Should that be done in the scope of this PR or separately?~~ Done in this PR.

[@sly7-7]: Just to clean issues when that awesome work is merged: 
Closes #2551, Fixes #2385, Fixes #2508, Closes #1491 

Fixes #2750, Closes #2739

rwjblue commented on packages/ember-data/lib/system/snapshot.js:
```diff
@@ -0,0 +1,299 @@
+
+var get = Ember.get;
+
+var hasPropertyAccessors = Ember.__loader.require('ember-metal/platform')['hasPropertyAccessors'];
```
This are exported by Ember as `Ember.platform.hasPropertyAccessors ` since at least 1.8.

https://github.com/emberjs/ember.js/blob/v1.8.1/packages/ember-metal/lib/main.js#L183-L186

rwjblue commented on packages/ember-data/lib/system/snapshot.js:
```diff
@@ -0,0 +1,299 @@
+var get = Ember.get;
+
+var hasPropertyAccessors = Ember.__loader.require('ember-metal/platform')['hasPropertyAccessors'];
+var defineProperty = Ember.__loader.require('ember-metal/properties')['defineProperty'];
```
This is exported as `Ember.defineProperty` (for a long time) and also as `Ember.platform.defineProperty` since at least 1.8.

https://github.com/emberjs/ember.js/blob/v1.8.1/packages/ember-metal/lib/main.js#L183-L186

rwjblue commented on packages/ember-data/lib/system/snapshot.js:
```diff
@@ -0,0 +1,299 @@
+    @deprecated Use [attr](#method_attr), [belongsTo](#method_belongsTo) or [hasMany](#method_hasMany) instead
+  */
+  get: function(keyName) {
+    Ember.deprecate('Using DS.Snapshot.get() is deprecated. Use .attr(), .belongsTo() or .hasMany() instead.');
```
This seems like a completely new API, why would we come out of the gates with deprecations?

Pardon me if this has already been discussed...

wecc replied:
I think @igorT, @wycats and @tomdale have discussed this.

igorT replied:
@rwjblue before when serializing something you would get a record and call `.get()` to access the properties. Now you will receive a snapshot and will call `.attr()`, `.belongsTo()`, `.hasMany()`. We still want to implement a depracated support for get in order to not break everyone's serializer

rwjblue replied:
@igorT - I see, thanks for the clarification.

wecc commented on packages/ember-data/lib/system/snapshot.js:
```diff
@@ -0,0 +1,283 @@
+          causes the internals of Chrome to access the constructor, which then
+          causes an infinite loop if accessed and calls `Ember.deprecate()`
+          again.
+        */
```
Ideas on how to improve this?

wecc commented:
Add test to make sure relationships aren't fetched.

bmac commented on packages/ember-data/lib/system/snapshot.js:
```diff
@@ -0,0 +1,298 @@
+      }
+
+      members = get(relationship, 'members');
+      members.forEach(function(member) {
```
Don't create the snapshots if we only want the ids

wecc replied:
My reasoning for this is to cache the hasMany on first access, no matter if you want just the IDs or the snapshots.

Imagine this:

```javascript
var snapshot = post.snapshot();
var ids = snapshot.hasMany('comments', { ids: true });
var comments = snapshot.hasMany('comments');
```

If we don't create the snapshots and cache them on the first access we could have a scenario when `ids` and `comments` in the example above doesn't match.

Thoughts?

igorT replied:
You wouldn't really be serializing twice while changing the data inbetween and the main reason we added `ids:true` was for perf so you don't have to create snapshots right?

bmac commented on packages/ember-data/lib/system/snapshot.js:
```diff
@@ -0,0 +1,298 @@
+    @param {string} keyName
+    @return {DS.Snapshot|undefined} A snapshot of a belongsTo relationship or undefined
+  */
+  belongsTo: function(keyName) {
```
Should this also support `snapshot.belongsTo(attr, { ids: true })` like `hasMany`?

wecc replied:
Yeah, that might be useful. There's one place that could benefit from this, see https://github.com/emberjs/data/pull/2623/files#diff-874af7e7df6c1be12e3a54caad8d859dR583

However, `ids` feels wrong to me since it's just one id? But `id` would be inconsistent to `hasMany`.

@igorT ?

bmac commented on packages/ember-data/lib/system/snapshot.js:
```diff
@@ -0,0 +1,298 @@
+
+  this.record = record;
+  this.type = record.constructor;
+  this.typeKey = record.constructor.typeKey;
```
We should have docs for these properties.

wecc replied:
Done!

bmac commented on packages/ember-data/lib/system/snapshot.js:
```diff
@@ -0,0 +1,298 @@
+/**
+  @class Snapshot
+  @namespace DS
+  @private
```
I don't think this needs to be marked private.

wecc replied:
I didn't remove `@private` to make sure constructor is hidden in API docs.

wecc commented:
- Added docs for `record`, `type` and `typeKey`
- Added `{ id: true }` option to `snapshot.belongsTo()` + test
- Snapshots are never created if passing `id/ids: true` to `snapshot.belongsTo()/hasMany()`, IDs gets cached separately

wycats commented:
@wecc you are a wonderful human being. Thanks so much for this!

wecc commented:
Merge conflicts fixed.

knownasilya commented:
So what exactly is a snapshot, and how would you use it?

wecc commented:
@knownasilya a Snapshot is a lighter representation of a record and it's current attributes and relationships. It's initial purpose is to be passed to serializers instead of the real records. This helps for example when serializers tries to get the current value of an async relationship, instead of fetching it (record behavior) a snapshot returns the current value regardless if it's loaded of not.

tomdale commented:
@knownasilya As @wecc described, the high order bit for me is that a snapshot represents the frozen state of a record at a particular moment in time. As we push everything towards being all-async all the time, you want to get a way to examine the current state of that record in the moment without triggering all sorts of side-effects, like loading relationships.

knownasilya commented:
Thanks for the clarification @wecc and @tomdale. It sure sounds like a useful and needed API, especially in the async scenario. I've struggled numerous times trying to get state without triggering any observers/cps in complicated scenarios, and it's a battle.

sly7-7 commented:
@tomdale @wecc From what I understand from the code, I think there are still inconsistencies in the implementation. Since the relationships are lazy and the attributes are eager, actually when you create a snapshot, and the access its data, I think problems can occur if the relationships of the underlying are changed before the snapshot's ones are accessed. Though I just may be completely wrong.

tomdale commented:
@sly7-7 You're absolutely correct. We discussed this at length and decided that you need to pre-emptively access the snapshot's relationships if you want to freeze them at that moment in time. If we eagerly snapshot all relationships as well as attributes, we may have to create thousands of snapshots that the user never needs. Definitely open to other suggestions here, because it's going to trip people up. But I don't know how to avoid making it easy to snapshot the entire object graph otherwise.

bakura10 commented:
I'm not sure to completely understand this feature. Can it be used to create "forked models" (à la EPF), so hat modifications are applied locally, and then saved independently (to avoid the very famous binding problem where properties are updated visually while the model is not saved yet)

tomdale commented:
@bakura10 It is not that complex. It simply, at snapshot time, loops over all of the attributes of a model and copies their current values over into an immutable hash, giving you a "snapshot in time" of the state of that model.

thomasvanlankveld commented:
If this does what I think it will do, it's going to make me very very happy :)

bmac commented on packages/ember-data/lib/system/snapshot.js:
```diff
@@ -0,0 +1,364 @@
+    @method id
+    @return {String} The ID of the snapshot's record
+  */
+  id: function() {
```
I'm nervous because now if you do `record.id` it will return the id of the record. This could break serializers that are relying on this behavior.

wecc commented:
Rebased, merge conflicts fixed.

bmac commented on packages/ember-data/lib/system/snapshot.js:
````diff
@@ -0,0 +1,361 @@
+    ```
+
+    @method id
+    @return {String} The ID of the snapshot's record
````
```
 @property id
 @type {String}
```
